### ISSUE TITLE
Errors with Distributed Training

### ISSUE THREAD

**Issue by VashishtMadhavan** [google-research/morph-net#78]
Hi, I am trying to train a network with the `GammaFlopsRegularizer` by distributing it over 4 gpus on my machine. Im using the [horovod](https://github.com/horovod/horovod) library to train a simple convolutional autencoder following the script below:

```
import numpy as np
import os
import tensorflow as tf
import horovod.tensorflow as hvd

from morph_net.network_regularizers import flop_regularizer
from morph_net.network_regularizers import latency_regularizer
from morph_net.tools import structure_exporter

# Define the Network
def conv_ae(x_ph, is_training, scope, channels=[32, 64, 64], reuse=False):
    norm_params = {'is_training': is_training, 'scale': True, 'center': False}
    # Network Definition
    with tf.variable_scope(scope, reuse=reuse):
        with slim.arg_scope([slim.conv2d, slim.conv2d_transpose],
                      normalizer_fn=slim.batch_norm,
                      normalizer_params=norm_params,
                      weights_initializer=tf.truncated_normal_initializer(0.0, 0.01),
                      weights_regularizer=slim.l2_regularizer(0.0005)):
            conv1 = slim.conv2d(x_ph, num_outputs=channels[0], kernel_size=3, scope='conv1')
            pool1 = slim.max_pool2d(conv1, kernel_size=2, scope='pool1')
            conv2 = slim.conv2d(pool1, num_outputs=channels[1], kernel_size=3, scope='conv2')
            pool2 = slim.max_pool2d(conv2, kernel_size=2, scope='pool2')
            conv3 = slim.conv2d(pool2, num_outputs=channels[2], kernel_size=3, scope='conv3')

            # deconv part
            deconv1 = slim.conv2d_transpose(conv3, num_outputs=channels[2], kernel_size=3, stride=2, scope='conv_transpose1')
            deconv2 = slim.conv2d_transpose(deconv1, num_outputs=channels[1], kernel_size=3, stride=2, scope='conv_transpose2')
            out = slim.conv2d_transpose(deconv2, num_outputs=1, kernel_size=3, stride=1, normalizer_fn=None, normalizer_params=None, activation_fn=tf.nn.sigmoid, scope='conv_output')
    return out


# Arguments
total_steps = 10000
batch_size = 128
val_batch_size = 1000
learning_rate = 1e-3
outdir = 'ae_morph_saves/'
gamma_scale = 1e-3
reg_strength = 1e-8
channels = [64, 64, 64]

# Initialize horovod
hvd.init()

# Load MNIST Data
train_data, test_data = tf.keras.datasets.mnist.load_data()
X_train, X_test = train_data[0], test_data[0]
global_step = tf.train.get_or_create_global_step()

# Optimization
N, H, W = X_train.shape
X_train_ph = tf.placeholder(tf.float32, [None, H, W, 1])
pred = conv_ae(X_train_ph, True, scope='base', channels=channels, reuse=False)
loss = tf.reduce_mean(tf.reduce_sum(tf.squared_difference(X_train_ph, pred), axis=[1,2,3]), axis=0)
model_vars = tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES, scope='base')

# applying L1 loss on batch norm gamma
network_regularizer = flop_regularizer.GammaFlopsRegularizer(
		output_boundary=[pred.op],
		input_boundary=[X_train_ph.op],
		gamma_threshold=gamma_scale)
reg_loss = network_regularizer.get_regularization_term() * reg_strength
cost_op = network_regularizer.get_cost()

# setting up distributed optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=learning_rate * hvd.size())
optimizer = hvd.DistributedOptimizer(optimizer)

gradients = optimizer.compute_gradients(loss + reg_loss, var_list=model_vars)
train_op = optimizer.apply_gradients(gradients, global_step=global_step)
exporter = structure_exporter.StructureExporter(network_regularizer.op_regularizer_manager)

hooks = [
	hvd.BroadcastGlobalVariablesHook(0),
	tf.train.StopAtStepHook(last_step=total_steps // hvd.size()),
	tf.train.LoggingTensorHook(tensors={'step': global_step, 'loss': loss}, every_n_iter=10),
]

# pin GPU to be used to process local rank (one GPU per process)
config = tf.ConfigProto()
config.gpu_options.allow_growth = True
config.gpu_options.visible_device_list = str(hvd.local_rank())
outdir = outdir if hvd.rank() == 0 else None

# Training Loop
with tf.train.MonitoredTrainingSession(checkpoint_dir=outdir,
										hooks=hooks, 
										config=config) as mon_sess:
	while not mon_sess.should_stop():
		idx = np.random.choice(X_train.shape[0], batch_size, replace=False)
		x_t = np.expand_dims(X_train[idx], axis=-1)
		feed_dict = {X_train_ph: x_t}
		g_step_val = mon_sess.run(global_step, feed_dict=feed_dict)
		_, grad_vals, structure_exporter_tensors, t_loss, reg_cost = mon_sess.run(
			[train_op, gradients, exporter.tensors, loss, cost_op],
			feed_dict=feed_dict)

		print("Step: ", g_step_val)
		print("Training Loss: ", t_loss)
		print("Reg Cost: ", reg_cost)

		# exporting model to JSON
		if hvd.rank() == 0:
			exporter.populate_tensor_values(structure_exporter_tensors)
			exporter.create_file_and_save_alive_counts(outdir, g_step_val)
```

When trying to train the model using `mpirun -np 4 -localhost localhost:4 -bind-to none -map-by slot -genvlist LD_LIBRARY_PATH,PATH  python <script-above>` it starts to hang, suggesting that the gradients are out of sync. The error is shown below:
```
Fri May 31 19:02:24 2019[1,0]<stderr>:[2019-05-31 19:02:24.396204: W horovod/common/operations.cc:588] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. 
Fri May 31 19:02:24 2019[1,0]<stderr>:Stalled ops:DistributedAdamOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_3_0 [missing ranks: 0, 1, 3]
Fri May 31 19:02:24 2019[1,0]<stderr>:[2019-05-31 19:02:24.396257: W horovod/common/operations.cc:588] DistributedAdamOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_8_0 [missing ranks: 0, 1, 3]
```
If I remove the `GammaFlopsRegularizer` portion, the error disappears and the model trains fine. My guess is that the regularizer is doing some non-deterministic pruning of connections. There seem to be some nodes where the corresponding gradient has been removed from the graph. I may be wrong, but is there a way to enforce this pruning to be deterministic across the distributed workers.

If not, is there another fix? I think this would be a good thing to look into as most large models that can be pruned rely on some form of distributed training.

**Comment by VashishtMadhavan** [comment]
It also seems that each time I initialize the network with morph net, the nodes in the graph are different, especially the gradient ops. This occurs before training too. 

I add the following bit before the training loop:
```
with open('nodes.txt', 'w') as f:
	nodes = [n.name for n in tf.get_default_graph().as_graph_def().node]
	nodes.sort()
	for i, node in enumerate(nodes):
		f.write('{}\n'.format(node))
```
This basically sorts the nodes of the TF graph and outputs them to a file. If I call `compute_gradients` on only `loss` and not `loss + reg_loss`, I get identical files for two seperate runs of the script. However, when I do compute gradients for `loss + reg_loss` the `nodes.txt` files are very different. Below is a short snippet of the diff:

```
1460a1476,1480
> gradients/mul_5_grad/Mul
> gradients/mul_5_grad/Mul_1
> gradients/mul_5_grad/tuple/control_dependency
> gradients/mul_5_grad/tuple/control_dependency_1
> gradients/mul_5_grad/tuple/group_deps
1496,1515d1515
< gradients/mul_6_grad/Mul
< gradients/mul_6_grad/Mul_1
< gradients/mul_6_grad/tuple/control_dependency
< gradients/mul_6_grad/tuple/control_dependency_1
< gradients/mul_6_grad/tuple/group_deps
< gradients/mul_7_grad/Mul
< gradients/mul_7_grad/Mul_1
< gradients/mul_7_grad/tuple/control_dependency
< gradients/mul_7_grad/tuple/control_dependency_1
< gradients/mul_7_grad/tuple/group_deps
< gradients/mul_8_grad/Mul
< gradients/mul_8_grad/Mul_1
< gradients/mul_8_grad/tuple/control_dependency
< gradients/mul_8_grad/tuple/control_dependency_1
< gradients/mul_8_grad/tuple/group_deps
< gradients/mul_9_grad/Mul
< gradients/mul_9_grad/Mul_1
< gradients/mul_9_grad/tuple/control_dependency
< gradients/mul_9_grad/tuple/control_dependency_1
< gradients/mul_9_grad/tuple/group_deps
```
I was wondering why the gradient ops in two identical networks, pre-training, are different.

**Comment by eladeban** [comment]
Hi,

Sorry for the delay in reply.

I am not sure I understand your question fully. Specifically, I am not familiar with horovod, but I can try to help.

1. Can you try to explicitly add `'fused'=True` to `norm_params = {'is_training': is_training, 'scale': True, 'center': False}`
2. Can you try **not to** use horovod and see if you get results this way. Note that the structure learning phase is often shorter so possibly by just waiting a bit, you won't have to _solve_ this problem. Later for retraining without the regularizer, you can turn on horovod back of course. 
3. **re gradients hanging**: 
 Can you try to use `GammaActivationRegularizer` it is a bit simpler than the FLPOs and it will be good to know if it works or not. FLOPregularizers are not actually continuous (due to the threshold) not sure if its a problem or not.
4. Another sanity check would be to just graph all variables ending with gamma and set
`reg = tf.add_n([tf.abs(var) for var in gammas])`
and see what happens.

5. **re gradients being different**:
If you  comparing the grads of `loss + reg_loss` and `loss` then they should be different, as reg_loss has gradients as well? or am I missing the points?

**Comment by VashishtMadhavan** [comment]
Thanks for the response and the help. Let me try to clear up some of the confusion:

1. I've added `'fused'=True` and changed the regularizer to the `GammaActivationRegularizer` and that seems to work with horovod, distributed over 4 gpus. However, distributed training still fails for the `GammaFlopsRegularizer`

2. The script works fine without horovod. It's only when I distribute training that the error shows up. I could try to learn the structure without horovod first, but this script is a proxy for a similar issue I have with a much larger network and that would take days to train if the training was not distributed.

3. **Re gradients being different: ** I'm comparing the graphs of two separate runs of network training using the script above, both of which optimize `loss + reg_loss`. Somehow the graphs for different runs contain different nodes even though the training script is the exact same. However, when I try training two runs that only optimize for `loss`, the graphs of the two network training runs are identical. These graphs are also logged before any training occurs, which precludes the possibility that graph modification by the network regularizer during training is causing the difference.

**Comment by VashishtMadhavan** [comment]
@eladeban any updates on this front?
